tower-retry/src/lib.rs

```diff
@@ -0,0 +1,138 @@
+#[macro_use]
+extern crate futures;
+extern crate tower_service;
+
+use futures::{Async, Future, Poll};
+use tower_service::Service;
+
+#[derive(Clone, Debug)]
+pub struct Retry<P, S> {
+    policy: P,
+    service: S,
+}
+
+#[derive(Debug)]
+pub struct ResponseFuture<P: Policy<S::Request, S::Response, S::Error>, S: Service> {
+    request: Option<S::Request>,
+    retry: Retry<P, S>,
+    state: State<S::Future, P::Future, S::Response, S::Error>,
+}
+
+#[derive(Debug)]
+enum State<F, P, R, E> {
+    /// Polling the future from `Service::call`
+    Called(F),
+    /// Polling the future from `Policy::retry`
+    Checking(P, Option<Result<R, E>>),
+    /// Polling `Service::poll_ready` after `Checking` was OK.
+    Retrying,
+}
+
+pub trait Policy<Req, Res, E>: Sized {
+    type Future: Future<Item=Self, Error=()>;
+    fn retry(&self, req: &Req, res: Result<&Res, &E>) -> Option<Self::Future>;
```

Comment:
My initial reaction to this was that I found it odd. Specifically, `retry` returning a future of `Self`.

Could you elaborate a bit more on the thought process that went into this design? I assume it has something to do w/ backoff.

Also, given that it takes `&self` and returns a future of `Self`, that implies cloning `Self` at some point. This could be tricky if `Self` contains data that cannot easily be cloned. Do you think this is a problem in practice?

Reply:
Yea, it was around wanting to provide backoff functionality. I realized that we could classify "successes" as needing a retry (like an HTTP 503), but an underlying `Backoff` wouldn't have known it was an error and that it should backoff.

> Also, given that it takes &self and returns a future of Self, that implies cloning Self at some point. 

It *could* mean cloning, or it could mean constructing a new version of `Self`. If there is expensive values to clone, they can be put in an `Arc`.

What I found useful here though is that it allows a retry policy to adapt depending on multiple factors: request, response or error, and it's current state. The `Limit` test case shows how it can therefore adapt, and build more "scoped" policies as retries go on.

Both of these combined are from noticing that finagle's `RetryPolicy` returns `Option<(Duration, RetryPolicy)>`. The duration allows delaying the retry, and the new policy can contain knowledge of the previous attempts. By returning a `Future` directly, it seemed more flexible. It *can* just be a delay of some kind, or it can be something else. And it means `tower_retry` doesn't need a dependency on `tokio_timer`.